### PR TITLE
DP-768: Store sessions in Redis

### PR DESCRIPTION
**Pull request by mareg**
To test it locally, add the following to your `compose.override.yaml` file:
```yaml
  organisation-app:
    container_name: ""
    deploy:
      replicas: 10
```
This will start with 10 replicas for the `organisation-app` and `gateway` will send traffic equally between them.

### CODE REVIEW

File: Frontend/CO.CDP.OrganisationApp/Program.cs [hunk 1]
```diff
@@ -61,7 +61,14 @@
 
 builder.ConfigureForwardedHeaders();
 
-builder.Services.AddDistributedMemoryCache();
+builder.Services.AddStackExchangeRedisCache(options =>
+{
+    var redisHost = builder.Configuration.GetValue<string>("Aws:ElastiCache:Hostname");
+    var redisPort = builder.Configuration.GetValue<string>("Aws:ElastiCache:Port");
```
Comment (jakzal) [inline Frontend/CO.CDP.OrganisationApp/Program.cs:67]: Since this is part of Aws configuration, is it worth to make it explicitly part of the `AwsConfiguration` config class? 

https://github.com/cabinetoffice/GCGS-Central-Digital-Platform/blob/main/Libraries/CO.CDP.AwsServices/AwsConfiguration.cs#L5-L14

With other Aws services we were also adding extension methods for convinnience, for example: 

https://github.com/cabinetoffice/GCGS-Central-Digital-Platform/blob/6d5cff1853759835f03dc974a8910e85c4d07f5f/Frontend/CO.CDP.OrganisationApp/Program.cs#L215C1-L220C50

File: compose.yml [hunk 2]
```diff
@@ -106,12 +115,15 @@ services:
         Organisation__Authority: '${Organisation__Authority:-http://gateway:${CDP_AUTHORITY_PORT:-8092}}'
 
   organisation-app:
-    container_name: co-cdp-organisation-app
+    # container_name: co-cdp-organisation-app
```
Comment (jakzal) [inline compose.yml:118]: Why this change?

**Review by andymantell — APPROVED** [review]

File: Frontend/CO.CDP.OrganisationApp/appsettings.json [hunk 3]
```diff
@@ -55,5 +63,7 @@
   "CharityCommission": {
     "Url": "",
     "SubscriptionKey": ""
-  }
+  },
+  "OneLoginLogoutNotificationUrls": [
```
Comment (dharmverma) [inline Frontend/CO.CDP.OrganisationApp/appsettings.json:67]: Looks like result of merge conflict, `OneLoginLogoutNotificationUrls `is not required to be added

File: Libraries/CO.CDP.AwsServices/ElastiCacheExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            Console.WriteLine("No elasti cache service configured.");
+            return services;
+        }
+
+        Console.WriteLine($"Elasti Cache Options: {awsConfiguration.ElastiCache}");
+        return services.AddStackExchangeRedisCache(options =>
+        {
+            var redisPort = awsConfiguration.ElastiCache.Port;
+            var redisHost = awsConfiguration.ElastiCache.Hostname;
+            Console.WriteLine($"Elasti Cache Options: {redisHost}:{redisPort}");
```
Comment (dharmverma) [inline Libraries/CO.CDP.AwsServices/ElastiCacheExtensions.cs:48]: Remove console log

File: Frontend/CO.CDP.OrganisationApp/Program.cs [hunk 5]
```diff
@@ -206,16 +212,11 @@
...
 builder.Services.AddHealthChecks();
-builder.Services
-    .AddAwsConfiguration(builder.Configuration)
-    .AddAwsS3Service()
-    .AddLoggingConfiguration(builder.Configuration)
-    .AddAmazonCloudWatchLogsService()
-    .AddCloudWatchSerilog(builder.Configuration);
 
 // @see DP-723 for details: https://noticingsystem.atlassian.net/browse/DP-723?focusedCommentId=27796
-// builder.Services.AddDataProtection()
```
Comment (dharmverma) [inline Frontend/CO.CDP.OrganisationApp/Program.cs:217]: Remove redundant comments

File: Libraries/CO.CDP.AwsServices/ElastiCacheExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            options.Configuration = $"{redisHost}:{redisPort}";
+            options.InstanceName = "Sessions";
```
Comment (dharmverma) [inline Libraries/CO.CDP.AwsServices/ElastiCacheExtensions.cs:50]: I would not name it session, because it just the cache, which is used by app level cache and user session both.
It is the .net, how its make use of it when managing session.
I would prefer to name something like "Cache" or "SirsiCache"
